### PR TITLE
[JENKINS-71182] Correct Unicode behavior of `XML_1_1`

### PR DESCRIPTION
**Pull request by jglick**
See [JENKINS-71182](https://issues.jenkins.io/browse/JENKINS-71182). ~There does not seem to be any good solution, until `PrettyPrintWriter` supports Unicode properly.~ https://github.com/jenkinsci/jenkins/pull/7875#discussion_r1183099044

### Testing done

See test cases. (Writing emojis was apparently not tested before, only reading.)

### Proposed changelog entries

- Fixed writing of emojis to XML (regression in 2.403).

### Proposed upgrade guidelines

N/A

### Maintainer checklist

Before the changes are marked as `ready-for-merge`:

- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/7924"></a>

### CODE REVIEW

**Comment by jglick** [comment]
(Oh, and I tried to use `WriterWrapper` to add indentation to `StaxWriter`, without success—the `>` at the end of an element is written only after text content or an end element is encountered.)

**Comment by basil** [comment]
I filed https://github.com/x-stream/xstream/pull/337 to fix XStream. Do you think it is worth temporarily inlining that patch into Jenkins while waiting for upstream to merge/release it, or do you think it is preferable to revert back to quirks mode?

**Comment by Bananeweizen** [comment]
I'm not sure if this is of any help here, but I have seen filtered streams/readers in other real world applications just stripping offending characters from the input stream before passing it to the XML handling (e.g. the NUL character that started these changes). Examples of software I have actually been using:
* https://josm.openstreetmap.de/browser/josm/trunk/src/org/openstreetmap/josm/io/InvalidXmlCharacterFilter.java?rev=6070
* https://github.com/cgeo/cgeo/blob/3eeb4b709665adc66a50ef77259e180c1040c5ce/main/src/main/java/cgeo/geocaching/files/InvalidXMLCharacterFilterReader.java

**Comment by jglick** [comment]
> Do you think it is worth temporarily inlining that patch into Jenkins while waiting for upstream to merge/release it, or do you think it is preferable to revert back to quirks mode?

Seems simplest to me to just use quirks mode until a fix makes it into a release. The fail-fast behavior is nice to have but hardly essential, assuming the `junit` update is adopted. No strong opinion.

> filtered streams/readers […] just stripping offending characters

Yes, and/or conversely rejecting `&#x0;` on output. Would add a little overhead. Again seems not worth the bother especially if a fixed XStream comes reasonably soon, but if someone wants to propose such a patch I will happily close this PR in favor of it.

**Comment by basil** [comment]
My preference would be to inline the XStream fix, since that would provide additional real-world testing which would increase the likelihood that the fix would be accepted upstream.

**Comment by jglick** [comment]
OK, I should be able to do that today.

File: core/src/main/java/hudson/util/PrettyPrintWriter.java
Comment (jglick) [inline core/src/main/java/hudson/util/PrettyPrintWriter.java:214]: This and matching lines are the fix from the upstream PR.

File: core/src/main/java/hudson/util/PrettyPrintWriter.java
Comment (jglick) [inline core/src/main/java/hudson/util/PrettyPrintWriter.java:62]: had to drop type parameter

File: core/src/main/java/hudson/util/PrettyPrintWriter.java
Comment (jglick) [inline core/src/main/java/hudson/util/PrettyPrintWriter.java:182]: had to switch to rawtypes

File: core/src/main/java/hudson/util/PrettyPrintWriter.java
Comment (jglick) [inline core/src/main/java/hudson/util/PrettyPrintWriter.java:15]: edited from original, obviously

File: core/src/main/java/hudson/util/PrettyPrintWriter.java
Comment (jglick) [inline core/src/main/java/hudson/util/PrettyPrintWriter.java:1]: Reformatting for Checkstyle, plus a couple minor changes noted inline.

File: core/src/main/java/hudson/util/PrettyPrintWriter.java
Comment (basil) [inline core/src/main/java/hudson/util/PrettyPrintWriter.java:55]: The last time I did something similar in #7270 I marked the class as `@Restricted(NoExternalUse.class)` since this is not intended to be a permanent API. I think it would make sense to do so here as well for the same reason.

**Review by basil — APPROVED** [review]

**Review by basil — APPROVED** [review]

File: core/src/main/java/hudson/util/PrettyPrintWriter.java
Comment (jglick) [inline core/src/main/java/hudson/util/PrettyPrintWriter.java:55]: dropped `public` on class and constructors

**Review by NotMyFault — APPROVED** [review]
Thank you, Jesse!

**Comment by basil** [comment]
This PR is now ready for merge. We will merge it after approximately 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process. Thanks!

**Pull request merged**
